File: backend/src/main/java/com/trio/backend/service/hr/EmployeeService.java
package com.trio.backend.service.hr;

import com.trio.backend.dto.hr.CreateEmployeeRequest;
import com.trio.backend.dto.hr.EmployeeResponse;
import com.trio.backend.dto.hr.EmployeeSearchCriteria;
import com.trio.backend.dto.hr.EmployeeStatistics;
import com.trio.backend.dto.hr.EmployeeTimelineEntry;
import com.trio.backend.dto.hr.UpdateEmployeeRequest;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;

import java.util.List;
import java.util.UUID;

public interface EmployeeService {

    EmployeeResponse create(UUID workspaceId, UUID departmentId, CreateEmployeeRequest request);

    EmployeeResponse getById(UUID workspaceId, UUID departmentId, UUID employeeId);

    Page<EmployeeResponse> list(UUID workspaceId, UUID departmentId, EmployeeSearchCriteria criteria, Pageable pageable);

    EmployeeResponse update(UUID workspaceId, UUID departmentId, UUID employeeId, UpdateEmployeeRequest request);

    void delete(UUID workspaceId, UUID departmentId, UUID employeeId);

    List<EmployeeTimelineEntry> getTimeline(UUID workspaceId, UUID departmentId, UUID employeeId);

    EmployeeStatistics getStatistics(UUID workspaceId, UUID departmentId);
}

File: frontend/project/src/services/employee-service.ts
import { apiClient } from '../lib/api';
import type { PageResponse } from '../types/api';

export interface EmployeeResponse {
  id: string;
  employeeNumber: string;
  firstName: string;
  lastName: string;
  email: string;
  phone?: string;
  address?: string;
  dateOfBirth?: string;
  nationality?: string;
  emergencyContact?: string;
  position: string;
  departmentId: string;
  teamId?: string;
  managerId?: string;
  employmentType: string;
  employmentStatus: string;
  startDate: string;
  endDate?: string;
  profilePicturePath?: string;
  candidateId?: string;
  createdAt: string;
  updatedAt: string;
}

export interface CreateEmployeeRequest {
  candidateId?: string;
  firstName: string;
  lastName: string;
  email: string;
  phone?: string;
  address?: string;
  dateOfBirth?: string;
  nationality?: string;
  emergencyContact?: string;
  position: string;
  teamId?: string;
  managerId?: string;
  employmentType: string;
  startDate: string;
}

export interface UpdateEmployeeRequest {
  firstName?: string;
  lastName?: string;
  email?: string;
  phone?: string;
  address?: string;
  dateOfBirth?: string;
  nationality?: string;
  emergencyContact?: string;
  position?: string;
  teamId?: string;
  managerId?: string;
  employmentType?: string;
  employmentStatus?: string;
  startDate?: string;
  endDate?: string;
  profilePicturePath?: string;
}

export interface EmployeeStatistics {
  totalEmployees: number;
  activeEmployees: number;
  onLeaveCount: number;
  probationCount: number;
  employeesByDepartment: Record<string, number>;
  employeesByTeam: Record<string, number>;
  employeesByEmploymentType: Record<string, number>;
  employeesByStatus: Record<string, number>;
  newHiresThisMonth: number;
}

export interface EmployeeTimelineEntry {
  id: string;
  eventType: string;
  title: string;
  description?: string;
  occurredAt: string;
  actorId?: string;
  actorName?: string;
}

function base(wsId: string, deptId: string) {
  return `/workspaces/${wsId}/departments/${deptId}/employees`;
}

export const employeeService = {
  list: (wsId: string, deptId: string, params?: { page?: number; size?: number; keyword?: string; status?: string }) =>
    apiClient.get<PageResponse<EmployeeResponse>>(`${base(wsId, deptId)}`, { params }),

  getById: (wsId: string, deptId: string, employeeId: string) =>
    apiClient.get<EmployeeResponse>(`${base(wsId, deptId)}/${employeeId}`),

  create: (wsId: string, deptId: string, data: CreateEmployeeRequest) =>
    apiClient.post<EmployeeResponse>(`${base(wsId, deptId)}`, data),

  update: (wsId: string, deptId: string, employeeId: string, data: UpdateEmployeeRequest) =>
    apiClient.put<EmployeeResponse>(`${base(wsId, deptId)}/${employeeId}`, data),

  delete: (wsId: string, deptId: string, employeeId: string) =>
    apiClient.delete<void>(`${base(wsId, deptId)}/${employeeId}`),

  getTimeline: (wsId: string, deptId: string, employeeId: string) =>
    apiClient.get<EmployeeTimelineEntry[]>(`${base(wsId, deptId)}/${employeeId}/timeline`),

  getStats: (wsId: string, deptId: string) =>
    apiClient.get<EmployeeStatistics>(`${base(wsId, deptId)}/stats`),
};
